### PR TITLE
Erase collection of non-consecutive nodes from vectors (issue #340)

### PR DESCRIPTION
**Pull request by cattabiani**
useful for simplifying https://github.com/BlueBrain/nmodl/pull/230

### CODE REVIEW

File: src/language/nodes.py [hunk 1]
```diff
@@ -272,6 +272,27 @@ def get_add_methods(self):
...
+                          */
+                         void erase_{to_snake_case(self.class_name)}(const std::unordered_set<{self.class_name}*>& to_be_erased) {{
+                            // loosely following the cpp reference of remove_if
+
+                            auto first = {self.varname}.begin();
+                            auto last = {self.varname}.end();
+                            auto result = first;
+                        
+                            while (first != last) {{
+                                if (to_be_erased.find(first->get()) == to_be_erased.end()) {{
```
Comment (ohm314) [inline src/language/nodes.py]: `<nitpick>`Unless I'm mistaken what you do here is move all the elements out of the way that should be kept, and then you delete the rest. Wouldn't be more efficient to do the opposite? Only move elements that should be deleted?

File: src/visitors/kinetic_block_visitor.hpp [hunk 2]
```diff
@@ -14,8 +14,8 @@
 
 #include <map>
 #include <memory>
-#include <set>
 #include <string>
+#include <unordered_set>
```
Comment (ohm314) [inline src/visitors/kinetic_block_visitor.hpp:18]: why the switch?

Comment (cattabiani) [reply src/visitors/kinetic_block_visitor.hpp:18]: There was no reason to have a set. Unordered sets perform better for finding nodes (required in the routines even before my addition)

File: src/language/nodes.py [hunk 3]
```diff
@@ -272,6 +272,27 @@ def get_add_methods(self):
                             auto last_it = const_iter_cast({self.varname}, last);
                             return {self.varname}.erase(first_it, last_it);
                          }}
+                         /**
+                          * \\brief Erase non-consecutive members to {self.varname}
+                          */
+                         void erase_{to_snake_case(self.class_name)}(const std::unordered_set<{self.class_name}*>& to_be_erased) {{
```
[the remark this answers is not in the document]
Comment (cattabiani) [reply src/language/nodes.py]: it is because the ordering is not required anywhere and set.find performs better

File: src/language/nodes.py (hunk 1 above)
Comment (cattabiani) [reply src/language/nodes.py]: Example: I need to remove an element at the beginning of the vector. If I call std::erase(begin()), the code moves all the other elements at the beginning and then erases the last one. In fact [here](http://www.cplusplus.com/reference/vector/vector/erase/) is stated that the complexity is:

> Linear on the number of elements erased (destructions) plus the number of elements after the last element deleted (moving).

So, in any case I need to move the remaining elements at the beginning and then erase the back part. The algorithm that I use is very similar to the one stated for [remove_if](http://www.cplusplus.com/reference/algorithm/remove_if/) as explained in the comment

**Comment by iomaganaris** [comment]
I would feel a bit safer if we had a test specifically for this function and using as an example a `StatementBlock` but apart from this looks good to me

**Comment by cattabiani** [comment]
> I would feel a bit safer if we had a test specifically for this function and using as an example a StatementBlock but apart from this looks good to me

I can try

**Comment by cattabiani** [comment]
I checked and the routine is already tested plenty of times in the sympy visitor tests. In the various tests it erases one, all and some of the elements of the `statements` vector with the sympy visitor. Thus, I do not think a specific test is required.

**Pull request merged**